go/plugins/compat_oai/openai/openai_live_test.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Log("genkit initialized")
+
+	// Initialize the OpenAI plugin
+	apiKeyOption := option.WithAPIKey(apiKey)
+	oai := openai.OpenAI{
+		Opts: []option.RequestOption{apiKeyOption},
+	}
+	oai.Init(ctx, g)
+
+	// Define a tool for calculating gablorkens
+	gablorkenTool := genkit.DefineTool(g, "gablorken", "use when need to calculate a gablorken",
+		func(ctx *ai.ToolContext, input struct {
+			Value float64
+			Over  float64
+		},
+		) (float64, error) {
+			return math.Pow(input.Value, input.Over), nil
+		},
+	)
+
+	t.Log("openai plugin initialized")
+
+	t.Run("embedder", func(t *testing.T) {
+		embedder := oai.Embedder(g, "text-embedding-3-small")
+		res, err := ai.Embed(ctx, embedder, ai.WithEmbedText("yellow banana"))
+		if err != nil {
+			t.Fatal(err)
+		}
+		out := res.Embeddings[0].Embedding
+		// There's not a whole lot we can test about the result.
+		// Just do a few sanity checks.
+		if len(out) < 100 {
+			t.Errorf("embedding vector looks too short: len(out)=%d", len(out))
+		}
+		var normSquared float32
+		for _, x := range out {
+			normSquared += x * x
+		}
+		if normSquared < 0.9 || normSquared > 1.1 {
+			t.Errorf("embedding vector not unit length: %f", normSquared)
+		}
+	})
+
+	t.Run("basic completion", func(t *testing.T) {
+		t.Log("generating basic completion response")
+		resp, err := genkit.Generate(ctx, g,
+			ai.WithPromptText("What is the capital of France?"),
+		)
+		if err != nil {
+			t.Fatal("error generating basic completion response: ", err)
+		}
+		t.Logf("basic completion response: %+v", resp)
+
+		out := resp.Message.Content[0].Text
+		if !strings.Contains(strings.ToLower(out), "paris") {
+			t.Errorf("got %q, expecting it to contain 'Paris'", out)
+		}
+
+		// Verify usage statistics are present
+		if resp.Usage == nil || resp.Usage.TotalTokens == 0 {
+			t.Error("Expected non-zero usage statistics")
+		}
+	})
+
+	t.Run("streaming", func(t *testing.T) {
+		var streamedOutput string
+		chunks := 0
+
+		final, err := genkit.Generate(ctx, g,
+			ai.WithPromptText("Write a short paragraph about artificial intelligence."),
+			ai.WithStreaming(func(ctx context.Context, chunk *ai.ModelResponseChunk) error {
+				chunks++
+				for _, content := range chunk.Content {
+					streamedOutput += content.Text
+				}
+				return nil
+			}))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// Verify streaming worked
+		if chunks <= 1 {
+			t.Error("Expected multiple chunks for streaming")
+		}
+
+		// Verify final output matches streamed content
+		finalOutput := ""
+		for _, content := range final.Message.Content {
+			finalOutput += content.Text
+		}
+		if streamedOutput != finalOutput {
+			t.Errorf("Streaming output doesn't match final output\nStreamed: %s\nFinal: %s",
+				streamedOutput, finalOutput)
+		}
+
+		t.Logf("streaming response: %+v", finalOutput)
+	})
+
+	t.Run("tool usage with basic completion", func(t *testing.T) {
```

Comment:
This function calling test is flaky depending on the model. It seems to be related to math.Pow(), and I'll give it a try to fix it.

Reply:
I believe this is fixed already

Reply:
Sorry, I haven't done this yet. What I’ve found so far is that the math.Pow() function returns a value between 11 and 13, depending on how the model generates code and handles floating-point values internally. Since all tests are currently passing and the issue isn’t critical, I’d leave it as is for now and revise it if it comes up again.